### ISSUE TITLE
Exceptions starting DDS can't be caught without running in a zone

### ISSUE THREAD

**Issue by DanTup** [dart-lang/sdk#55731]
The following code tries to start DDS but passes an invalid URI:

```dart
import 'dart:async';

import 'package:dds/dds.dart';

Future<void> main() async {
  try {
    final a = await DartDevelopmentService.startDartDevelopmentService(
      Uri.parse('http://bogus.local/'), // This isn't valid
    );
    print('1');
  } catch (e) {
    print('2');
  }
  print('3');
}
```

None of the numbers here get printed, but the app terminates with an unhandled exception:

```
Unhandled exception:
SocketException: Failed host lookup: 'bogus.local' (OS Error: No such host is known.
, errno = 11001)
#0      _NativeSocket.startConnect (dart:io-patch/socket_patch.dart:721:35)
#1      _RawSocket.startConnect (dart:io-patch/socket_patch.dart:1914:26)
#2      RawSocket.startConnect (dart:io-patch/socket_patch.dart:27:23)
#3      Socket._startConnect (dart:io-patch/socket_patch.dart:2138:22)
#4      Socket.startConnect (dart:io/socket.dart:805:21)
#5      _ConnectionTarget.connect (dart:_http/http_impl.dart:2484:20)
#6      _HttpClient._getConnection.connect (dart:_http/http_impl.dart:2923:12)
#7      _HttpClient._getConnection (dart:_http/http_impl.dart:2928:12)
#8      _HttpClient._openUrl (dart:_http/http_impl.dart:2783:12)
#9      _HttpClient.openUrl (dart:_http/http_impl.dart:2620:7)
#10     _WebSocketImpl.connect (dart:_http/websocket_impl.dart:1021:42)
#11     WebSocket.connect (dart:_http/websocket.dart:320:22)
#12     new IOWebSocketChannel.connect (package:web_socket_channel/io.dart:81:28)
#13     connect (package:web_socket_channel/src/_connect_io.dart:15:24)
#14     new WebSocketChannel.connect (package:web_socket_channel/src/channel.dart:144:16)
#15     _defaultWebSocketBuilder (package:dds/src/dds_impl.dart:54:27)
#16     DartDevelopmentServiceImpl.startService (package:dds/src/dds_impl.dart:82:40)
#17     DartDevelopmentService.startDartDevelopmentService (package:dds/dds.dart:97:19)
#18     main (file:///C:/Dev/Test%20Projects/write_to_exited_process/bin/vm.dart:7:44)
#19     _delayEntrypointInvocation.<anonymous closure> (dart:isolate-patch/isolate_patch.dart:297:19)
#20     _RawReceivePort._handleMessage (dart:isolate-patch/isolate_patch.dart:184:12)

Exited (255).
```

Speaking to @bkonyi, it sounds like this is not intended and we shouldn't need to wrap our own zones (in this case in the DAP startup code that tries to ensure there is a DDS for the target process) to handle these errors.

**Comment by bkonyi** [comment]
@derekxu16 can you take a look when you have a chance?

**Comment by bkonyi** [comment]
> Possibly related to this (but if not I can file another issue)

This issue felt familiar, so I did some digging and found this: https://github.com/flutter/flutter/pull/141821

Basically, this is just the behavior of zones, which is kind of gross.

**Comment by DanTup** [comment]
Ah, thanks for the info, that explains what I see. I handled this by using a completer in https://dart-review.googlesource.com/c/sdk/+/366662 but that's also quite gross, so if we could eliminate the need for a zone (the original issue here) we can remove than and just have the original simple try/catch :-)

**Comment by derekxu16** [comment]
@bkonyi, so is our goal here to make exceptions thrown when starting DDS catchable? This `webSocketBuilder` call currently doesn't run in a `Zone` with an error handler (https://github.com/dart-lang/sdk/blob/fe91f3eed881db23f51dda65616b004a35ed2f57/pkg/dds/lib/src/dds_impl.dart#L83). So I think that if we want to make exceptions thrown when starting DDS catchable, we have to use a `Completer` + `runZonedGuarded` similarly to how @DanTup used them in his CL.

**Comment by bkonyi** [comment]
Yeah, basically we just don't want unhandled asynchronous exceptions to leak out of DDS. If it makes sense to throw (e.g., if `remoteVmServiceWsUri` points to a VM service that doesn't exist) we should throw synchronously. If it makes sense to swallow an exception (e.g., the VM service connection disappears, which indicates we DDS should shutdown), we should do that and shutdown gracefully.

**Comment by derekxu16** [comment]
I've just published `dds 4.2.3`, which includes the fix.

https://pub.dev/packages/dds/versions/4.2.3

**Comment by DanTup** [comment]
Thanks! I've opened https://dart-review.googlesource.com/c/sdk/+/368567 to remove the unnecessary zone (and confirmed the test still passes).
